src/main/java/org/openstreetmap/atlas/checks/validation/tag/TollValidationCheck.java

```diff
@@ -0,0 +1,624 @@
+        }
+        return Angle.NONE;
+    }
+
+    /**
+     * @param tags
+     *            some osm tags
+     * @return boolean if the barrier tag contains 'toll'
+     */
+    private boolean barrierTagContainsToll(final Map<String, String> tags)
+    {
+        return tags.get(BarrierTag.KEY).contains(TollTag.KEY);
+    }
+
+    /**
+     * @param tags
+     *            some edge tags
+     * @param tags2
+     *            some other edge tags
+     * @return both sets of tags have toll=yes
+     */
+    private boolean bothTollYesTag(final Map<String, String> tags, final Map<String, String> tags2)
+    {
+        return this.hasTollYesTag(tags) && this.hasTollYesTag(tags2);
+    }
+
+    /**
+     * @param tags
+     *            some osm tags
+     * @return boolean if tags contains key 'barrier'
+     */
+    private boolean containsBarrierTag(final Map<String, String> tags)
+    {
+        return tags.containsKey(BarrierTag.KEY);
+    }
+
+    /**
+     * @param tags
+     *            some osm tags
+     * @return boolean if tags contains key 'highway'
+     */
+    private boolean containsHighwayTag(final Map<String, String> tags)
+    {
+        return tags.containsKey(HighwayTag.KEY);
+    }
+
+    /**
+     * @param tags
+     *            some osm tags
+     * @return boolean if tags contains key 'toll'
+     */
+    private boolean containsTollTag(final Map<String, String> tags)
+    {
+        return tags.containsKey(TollTag.KEY);
+    }
+
+    /**
+     * @param edge
+     *            some edge
+     * @return boolean if edge intersects toll feature
+     */
+    private boolean edgeIntersectsTollFeature(final Edge edge)
+    {
+        final Iterable<Area> intersectingAreas = edge.getAtlas().areasIntersecting(edge.bounds());
+        final Iterable<Node> edgeNodes = edge.connectedNodes();
+        for (final Area area : intersectingAreas)
+        {
+            final boolean areaContainsPolyline = area.asPolygon().overlaps(edge.asPolyLine());
+            final Map<String, String> areaTags = area.getOsmTags();
+            if (!areaContainsPolyline)
+            {
+                continue;
+            }
+            return this.containsBarrierTag(areaTags) && this.barrierTagContainsToll(areaTags);
+        }
+
+        for (final Node node : edgeNodes)
+        {
+            final Map<String, String> nodeTags = node.getOsmTags();
+            if ((this.containsHighwayTag(nodeTags) && this.highwayTagContainsToll(nodeTags))
+                    || this.containsBarrierTag(nodeTags) && this.barrierTagContainsToll(nodeTags))
+            {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * @param edge
+     *            edge in question
+     * @param alreadyCheckedObjectIds
+     *            already been touched IDs
+     * @return edge that proves backwards is escapable
+     */
+    private Optional<Edge> edgeProvingBackwardsIsEscapable(final Edge edge,
```

Comment:
Could we combine this and `edgeProvingForwardIsEscapable()` into a new method where the params would just be `Edge edge, Set<Edge> connectedEdges, Set<Long> alreadyCheckedObjectIds)`?

Reply:
Actually I just noticed they call each other... if there's not already a stack overflow problem or anything I don't think my suggestion alone should cause that either

Reply:
I personally would like to keep them separate just for ease of access and understandability? Same with the nearby toll features request. If this is an issue maybe the PR could be accepted without combining the functions and I can see what I Can work out when I add in the autofix suggestions. What do you think?

Reply:
Yeah that sounds good. Usually sonar throws a duplication warning but looks like we're good here since inEdges is replaced by outEdges in a few lines